qemu/tests/virtio_fs_memory_leak_check.py

```diff
@@ -0,0 +1,94 @@
+from provider.storage_benchmark import generate_instance
+
+
+@error_context.context_aware
+def run(test, params, env):
+    """
+    Check if there is memory leak in the vm during io test on virtiofs.
+
+    1) Boot guest with virtiofs device.
+    2) Enable driver verifier first.
+    3) Remove the poolmon result file.
+    4) Record the non-paged pool Mmdi info.
+    5) Run iozone test on virtiofs.
+    6) Record the non-paged pool Mmdi info after iotest.
+    5) Check if there is Mmdi Diff increation.
+
+    :param test: QEMU test object
+    :param params: Dictionary with the test parameters
+    :param env: Dictionary with test environment.
+    """
+
+    def _record_poolmon():
+        """Record Mmdi tag memory info in pool monitor to C: volume"""
+        status = session.cmd_status(poolmon_mmdi_cmd)
+        if status:
+            test.fail("Fail to get Mmdi pool tag memory "
+                      "info in pool monitor.")
+
+    driver = params["driver_name"]
+    driver_verifier = params.get("driver_verifier", driver)
+    driver_running = params.get('driver_running', driver_verifier)
+    timeout = params.get_numeric("login_timeout", 360)
+    fs_target = params.get("fs_target")
+    test_file = params.get('virtio_fs_test_file')
+    iozone_options = params.get('iozone_options')
+    io_timeout = params.get_numeric("io_timeout", 1800)
+    poolmon_mmdi_cmd = params["poolmon_mmdi_cmd"]
+    get_mem_poolmon_cmd = params["get_mem_poolmon_cmd"]
+    record_file = params["record_file"]
+
+    vm_name = params['main_vm']
+    vm = env.get_vm(vm_name)
+    vm.verify_alive()
+    error_context.context("Boot guest with %s device" % driver, test.log.info)
+    session = vm.wait_for_login(timeout=timeout)
+
+    error_context.context("Run the viofs service", test.log.info)
+    utils_test.qemu.windrv_verify_running(session, test, driver_running)
+    session = utils_test.qemu.setup_win_driver_verifier(session,
+                                                        driver_verifier,
+                                                        vm)
+    virtio_fs_utils.run_viofs_service(test, params, session)
+
+    driver_letter = virtio_fs_utils.get_virtiofs_driver_letter(test,
+                                                               fs_target,
+                                                               session)
+    fs_dest = "%s:" % driver_letter
+    guest_file = os.path.join(fs_dest, test_file).replace("/", "\\")
+
+    test.log.info("Record memory info before iotest.")
+    poolmon_mmdi_cmd = utils_misc.set_winutils_letter(session,
+                                                      poolmon_mmdi_cmd)
+    if session.cmd_status("dir %s" % record_file) == 0:
+        test.log.info("Removing file %s." % record_file)
+        session.cmd_status("del /f /q %s" % record_file)
+    _record_poolmon()
+
+    error_context.context("Start iozone test.", test.log.info)
+    try:
+        io_test = generate_instance(params, vm, 'iozone')
```

Comment:
```suggestion
    io_test = generate_instance(params, vm, 'iozone')
    try:    
```
I will raise the `NameError: name 'io_test' is not defined` in the `finally` blocker if failed to generate the `io_test` instance.
So we suggest moving the instance creation out of the `try` blocker to ensure the exectption is correct.

Reply:
If the failure occured during generate instance phase, it means part of fio env were setup, it still needs to clear the env. I prefer to keep it in try condition like most of benchmark storage function did, what do you think?

Reply:
> If the failure occured during generate instance phase, it means part of fio env were setup, it still needs to clear the env. I prefer to keep it in try condition like most of benchmark storage function did, what do you think?

You may not worry about it since there is decorator function `@StorageBenchmark._clean_env` to clean up the env file for the generate instance phase, please refer to https://github.com/autotest/tp-qemu/blob/6989b15632b5325f7b7ddf6f5fde081373b06ae0/provider/storage_benchmark.py#L319

Reply:
Thanks, updated.